File: api/internal/reviewai/observation_test.go
package reviewai

import (
	"strings"
	"testing"
)

func TestRedactVisibleTextRemovesSecrets(t *testing.T) {
	input := "email y@example.com password hunter2 token abc123 sk-ant-secret"
	got := RedactVisibleText(input)
	for _, forbidden := range []string{"hunter2", "abc123", "sk-ant-secret"} {
		if strings.Contains(got, forbidden) {
			t.Fatalf("redacted text leaked %q: %s", forbidden, got)
		}
	}
}

func TestRedactDOMHintsDropsPasswordControls(t *testing.T) {
	hints := []DOMHint{
		{Role: "textbox", Text: "Password", SelectorHint: "input[type=password]"},
		{Role: "button", Text: "Connect TikTok", SelectorHint: "[data-review-step='connect-tiktok']"},
	}
	got := RedactDOMHints(hints)
	if len(got) != 1 || got[0].Text != "Connect TikTok" {
		t.Fatalf("unexpected hints after redaction: %+v", got)
	}
}

File: review-agent/tests/observation.test.js
import test from "node:test";
import assert from "node:assert/strict";

import { collectPageObservation, redactObservation } from "../src/observation.js";

test("redactObservation removes password-like visible text and DOM hints", () => {
  const obs = redactObservation({
    visible_text: "Password hunter2 token abc123 Connect TikTok",
    dom_hints: [
      { role: "textbox", text: "Password", selector_hint: "input[type=password]" },
      { role: "button", text: "Connect TikTok", selector_hint: "[data-review-step='connect-tiktok']" },
    ],
  });

  assert.equal(obs.visible_text.includes("hunter2"), false);
  assert.equal(obs.visible_text.includes("abc123"), false);
  assert.deepEqual(obs.dom_hints, [
    { role: "button", text: "Connect TikTok", selector_hint: "[data-review-step='connect-tiktok']" },
  ]);
});

test("collectPageObservation captures url title text and review-step hints", async () => {
  const page = {
    url: () => "https://review.example.com/tiktok/posting",
    title: async () => "TailTales",
    locator: () => ({ allInnerTexts: async () => ["Connect TikTok", "Upload video"] }),
    evaluate: async () => [
      { role: "button", text: "Connect TikTok", selector_hint: "[data-review-step='connect-tiktok']" },
    ],
  };

  const obs = await collectPageObservation(page, { jobId: "rvjob_1", stepKey: "connect_tiktok" });

  assert.equal(obs.job_id, "rvjob_1");
  assert.equal(obs.current_url, "https://review.example.com/tiktok/posting");
  assert.equal(obs.page_title, "TailTales");
  assert.equal(obs.visible_text, "Connect TikTok\nUpload video");
  assert.equal(obs.dom_hints[0].selector_hint, "[data-review-step='connect-tiktok']");
});
